rotkehlchen/tests/external_apis/test_cryptocompare.py

```diff
@@ -177,3 +184,73 @@ def test_cryptocompare_query_compound_tokens(
         to_asset=A_USD,
     )
     assert price is not None
+
+
+@pytest.mark.parametrize('from_asset, to_asset, timestamp, expected_price', [
+    (A_ETH, A_USD, Timestamp(1592629200), Price(ZERO)),
+    (A_COMP, A_COMP, Timestamp(1592629200), Price(ZERO)),  # both assets COMP
+    (A_USD, A_USD, Timestamp(1592629200), Price(ZERO)),  # both assets USD
+    (A_COMP, A_USDT, Timestamp(1592629200), Price(ZERO)),  # to_asset USDT
+    (A_USDT, A_COMP, Timestamp(1592629200), Price(ZERO)),  # from_asset USDT
+    (A_COMP, A_USD, Timestamp(1592629200), Price(FVal('239.13'))),
+    (A_USD, A_COMP, Timestamp(1592629200), Price(FVal('0.004181825785137791159620290219'))),
+    (A_COMP, A_USD, Timestamp(1592629201), Price(ZERO)),  # timestamp gt
+    (A_USD, A_COMP, Timestamp(1592629201), Price(ZERO)),  # timestamp gt
+])
+def test_check_and_get_special_histohour_price(
+        cryptocompare,
+        from_asset,
+        to_asset,
+        timestamp,
+        expected_price,
+):
+    """
+    Test expected prices are returned for different combinations of
+    `from_asset`, `to_asset` and `timestamp`.
+    """
+    price = cryptocompare._check_and_get_special_histohour_price(
+        from_asset=from_asset,
+        to_asset=to_asset,
+        timestamp=timestamp,
+    )
+    assert price == expected_price
+
+
+def test_keep_special_histohour_cases_up_to_date(cryptocompare):
+    """Test CRYPTOCOMPARE_SPECIAL_HISTOHOUR_CASES assets timestamps and prices
+    are still valid.
+    """
+    limit = 10
+    for asset, asset_data in CRYPTOCOMPARE_SPECIAL_HISTOHOUR_CASES.items():
+        # Call `query_endpoint_histohour()` for handling special assets
+        to_timestamp = Timestamp(asset_data.timestamp - 3600)
+        from_timestamp = Timestamp(to_timestamp - limit * 3600)
+        response = cryptocompare.query_endpoint_histohour(
+            from_asset=asset,
+            to_asset=A_USD,
+            limit=limit,
+            to_timestamp=to_timestamp,
+        )
+        if response['Data']:
+            for hour_price_data in response['Data']:
+                if (
+                    hour_price_data['high'] == 0 and
+                    hour_price_data['low'] == 0 and
+                    hour_price_data['open'] == 0 and
+                    hour_price_data['close'] == 0
+                ):
+                    break
+            else:
```

Comment:
This `else` is sitting under the `for` which is quite wrong. I am wondering how come no linter tells you anything here ...


Also why do an `if`, `else` here? You have a tendency to make things complicated. All you have to do here is an if check to see if anything non-zero is there at the timestamp and add a warning.

```python
if any([hour_price_data[x] for x in ('high', 'low', 'open', 'close')]):
    warning_msg = (
        f'Cryptocompare histohour API has non-zero prices for asset '
        f'{asset.identifier} from {from_timestamp} to {to_timestamp}. '
        f' Please, update CRYPTOCOMPARE_SPECIAL_HISTOHOUR_CASES dict.',
    )
    test_warnings.warn(UserWarning(warning_msg))
```

Reply:
`for...else` and `while...else` are [valid](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops) in python.

You think I have a tendency for overcomplicating things but I'm just coding just a bit safe given my limited experience of this API, how Rotki uses its attributes and Rotki in general.

Your code is not testing what mine is testing. I'm testing that any timestamp below the one in the dict, will contain in the response an entry with **all** the price attributes at zero. Your code will log a warning either if all the attributes have a price   greater than zero or if any attribute has a price greater than zero.

From what I've experienced using the API, the rows to be discarded are the ones with all price values at 0; this is the common pattern. Besides, and I may be wrong, couldn't a valid entry price have open/close/low at zero (e.g. exit scam, token crashing)? As I've seen that for timestamps that fall between hours `query_historical_price()` uses the low and high avg, I just wanted to play safe and assert all price attributes are at 0.

Regarding the `if..else`, it is just in case `response` has no results, which would make the test pass and mislead us. An assert wouldn't be fair as it is the API fault.

If my testing approach is wrong (not the implementation, but what I'm testing), let me know.

Reply:
Hey @victornsc The `else` after a while is an arcane construct that should never be used. It's really hard to read and understand. Please don't use it. I am not the only one who thinks [it's not a good idea](https://stackoverflow.com/a/9980160/110395).

My code, as I have told you before, is meant as pseudo-code guidance to show you how to simplify. I write it in the 2-5 mins I can afford while reviewing your PR.

> Besides, and I may be wrong, couldn't a valid entry price have open/close/low at zero (e.g. exit scam, token crashing)?

It shouldn't. This is just to check special cases where the price is already at zero. So we don't deal with any such problems.


The change you made is good thank you.

I will leave one comment and merge.